src/sparsebase/io/mtx_writer.cc

```diff
@@ -0,0 +1,320 @@
+      }
+      if (format_ == "array" && field_ == "pattern") 
+      {
+        throw utils::ReaderException("Matrix market files with array format cannot have the field ""'pattern' ");
+      }
+      if (format_ == "array" && symmetry_ != "general") 
+      {
+        throw utils::ReaderException("Matrix market files with array format cannot have the property 'symmetry' ");
+      }
+      if (symmetry_ == "hermitian") 
+      {
+        throw utils::ReaderException("Matrix market writer does not currently support hermitian symmetry.");
+      }
+
+      std::ofstream mtxFile;
+      mtxFile.open(filename_);
+      
+      //header
+      mtxFile << "%%MatrixMarket " << object_ << " " << format_ << " " << field_ << " " << symmetry_ << "\n";
+      
+
+      //TODO add warning when given field and actual data format is not same (integer vs float etc)
+
+      if constexpr (std::is_same_v<ValueType, void>)
+      {
+        throw utils::WriterException("Cannot write an MTX with void ValueType");
+      }
+      else
+      {
+        auto dimensions = coo->get_dimensions();
+        IDType* row = coo->get_row();
+        IDType* col = coo->get_col();
+        ValueType* val = coo->get_vals();
+
+        if (val != NULL && field_ == "pattern")
+        {
+          std::cerr << "Warning: Pattern selected but values given.\n";
+        }
+
+        //Symmetry check
+        bool saidSymmetric = (symmetry_ == "symmetric" || symmetry_ == "skew-symmetric" || symmetry_ == "hermitian");
+        int NNZ = coo->get_num_nnz();
+        int count_symmetric = 0;
+        int count_diagonal = 0;
+        if (saidSymmetric && dimensions[0] == dimensions[1]) {
+            for (int i = 0; i < coo->get_num_nnz(); ++i) {
+                if (row[i] != col[i]) { // Non-diagonal entry, check symmetric counterpart
+                    bool found_symmetric = false;
+                    for (int j = 0; j < coo->get_num_nnz(); ++j) {
+                      if (symmetry_ == "skew-symmetric")
+                      {
+                        if (row[j] == col[i] && col[j] == row[i] && val[j] == -val[i]) {
+                            found_symmetric = true;
+                            count_symmetric++;
+                            break;
+                        }
+                      }
+                      else
+                      {
+                        if (row[j] == col[i] && col[j] == row[i] && val[j] == val[i]) {
+                            found_symmetric = true;
+                            count_symmetric++;
+                            break;
+                        }
+                      }
+                        
+                    }
+                    if (!found_symmetric) {
+                        throw utils::ReaderException("Matrix is not symmetric!");
+                    }
+                }
+                else //diagonal
+                {
+                  count_diagonal++;
```

Comment:
Here, if the symmetry is skew-symmetric, then there is an exception should be thrown. This is because diagonal values should be 0 for skew-symmetric matrices.

Reply:
I added a check for diagonal entries if any of them is non zero and symmetry is skew symmetric, I throw an exception now.